File: vertx-service-discovery/src/test/resources/polyglot/my-verticle.js
var ServiceDiscovery = require("vertx-service-discovery-js/service_discovery");
var HttpEndpoint = require("vertx-service-discovery-js/http_endpoint");
var EventBusService = require("vertx-service-discovery-js/event_bus_service");
var MessageSource = require("vertx-service-discovery-js/message_source");
var DataSource = require("vertx-service-discovery-js/jdbc_data_source");
var RedisDataSource = require("vertx-service-discovery-js/redis_data_source");
var MongoDataSource = require("vertx-service-discovery-js/mongo_data_source");
var HelloService = require("test-services-js/hello_service");
var HttpClient = require("vertx-js/http_client");
var MessageConsumer = require("vertx-js/message_consumer");
var RedisClient = require("vertx-redis-js/redis_client");
var MongoClient = require("vertx-mongo-js/mongo_client");
var JDBCClient = require("vertx-jdbc-js/jdbc_client");

function getVertx() {
  return vertx;
}

var discovery = ServiceDiscovery.create(getVertx());

function getDelegate(obj) {
  return "" + obj._jdel;
}

getVertx().eventBus().consumer("http-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-http-service"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no http service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = getDelegate(reference);
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        var client = reference.getAs(HttpClient);
        if (!client) {
          message.reply("FAIL - client is null");
        } else {
          result.client_del = getDelegate(client);
          result.cached_del = getDelegate(reference.cachedAs(HttpClient));
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});

getVertx().eventBus().consumer("http-sugar", function (message) {
  HttpEndpoint.getClient(discovery, function (rec) {
    return rec.name === "my-http-service"
  }, function (res, err) {
    if (err) {
      message.reply("FAIL - no http service");
    } else {
      var result = {};
      if (!res) {
        message.reply("FAIL - client is null");
      } else {
        result.client_del = getDelegate(res);
        message.reply(result);
      }
    }
  });
});

getVertx().eventBus().consumer("service-sugar", function (message) {
  EventBusService.getServiceProxy(discovery, function (rec) {
      return rec.name === "my-service"
    },
    HelloService,
    function (res, err) {
      if (err) {
        message.reply("FAIL - no service");
      } else {
        var result = {};
        if (!res) {
          message.reply("FAIL - client is null");
        } else {
          result.client = res.toString();
          result.client_del = getDelegate(res);
          message.reply(result)
        }
      }
    });
});

getVertx().eventBus().consumer("service-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-service"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = reference._jdel.toString();
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        // Must create the object.
        var proxy = reference.getAs(HelloService);

        if (!proxy) {
          message.reply("FAIL - client is null");
        } else {
          result.client = proxy.toString();
          result.client_del = getDelegate(proxy);
          result.cached_del = getDelegate(reference.cachedAs(HelloService));
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});


getVertx().eventBus().consumer("ds-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-data-source"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no http service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = getDelegate(reference);
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        var client = reference.getAs(JDBCClient);

        if (!client) {
          message.reply("FAIL - client is null");
        } else {
          result.client_del = getDelegate(client);
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});

getVertx().eventBus().consumer("ds-sugar", function (message) {
  DataSource.getJDBCClient(discovery, function (rec) {
    return rec.name === "my-data-source"
  }, function (res, err) {
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var result = {};
      if (!res) {
        message.reply("FAIL - client is null");
      } else {
        result.client_del = getDelegate(res);
        message.reply(result);
      }
    }
  });
});



getVertx().eventBus().consumer("redis-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-redis-data-source"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = "" + reference._jdel;
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        var client = reference.getAs(RedisClient);
        if (!client) {
          message.reply("FAIL - client is null");
        } else {
          result.client_del = "" + client._jdel;
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});

getVertx().eventBus().consumer("redis-sugar", function (message) {
  RedisDataSource.getRedisClient(discovery, function (rec) {
    return rec.name === "my-redis-data-source"
  }, function (res, err) {
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var result = {};
      if (!res) {
        message.reply("FAIL - client is null");
      } else {
        result.client_del = getDelegate(res);
        message.reply(result);
      }
    }
  });
});

getVertx().eventBus().consumer("mongo-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-mongo-data-source"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = "" + reference._jdel;
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        var client = reference.getAs(MongoClient);
        if (!client) {
          message.reply("FAIL - client is null");
        } else {
          result.client_del = "" + client._jdel;
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});

getVertx().eventBus().consumer("mongo-sugar", function (message) {
  MongoDataSource.getMongoClient(discovery, function (rec) {
    return rec.name === "my-mongo-data-source"
  }, function (res, err) {
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var result = {};
      if (!res) {
        message.reply("FAIL - client is null");
      } else {
        result.client_del = getDelegate(res);
        message.reply(result);
      }
    }
  });
});

getVertx().eventBus().consumer("source1-ref", function (message) {
  discovery.getRecord(function (rec) {
    return rec.name === "my-message-source-1"
  }, function (rec, err) {
    var result = {};
    if (err) {
      message.reply("FAIL - no http service");
    } else {
      var reference = discovery.getReference(rec);
      result.ref_del = "" + reference._jdel;
      if (!reference) {
        message.reply("FAIL - reference is null");
      } else {
        var client = reference.getAs(MessageConsumer);

        if (!client) {
          message.reply("FAIL - client is null");
        } else {
          result.client_del = "" + client._jdel;
          message.reply(result);
          reference.release();
        }
      }
    }
  });
});

getVertx().eventBus().consumer("source1-sugar", function (message) {
  MessageSource.getConsumer(discovery, function (rec) {
    return rec.name === "my-message-source-1"
  }, function (res, err) {
    if (err) {
      message.reply("FAIL - no service");
    } else {
      var result = {};
      if (!res) {
        message.reply("FAIL - client is null");
      } else {
        result.client_del = res._jdel.toString();
        message.reply(result);
      }
    }
  });
});

File: vertx-service-discovery/src/test/java/io/vertx/servicediscovery/polyglot/MyVerticle.java
package io.vertx.servicediscovery.polyglot;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.eventbus.EventBus;
import io.vertx.core.eventbus.MessageConsumer;
import io.vertx.core.http.HttpClient;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.jdbc.JDBCClient;
import io.vertx.ext.mongo.MongoClient;
import io.vertx.redis.client.Redis;
import io.vertx.servicediscovery.ServiceDiscovery;
import io.vertx.servicediscovery.ServiceReference;
import io.vertx.servicediscovery.service.HelloService;
import io.vertx.servicediscovery.types.*;
import io.vertx.ext.web.client.WebClient;

/**
 * @author <a href="http://escoffier.me">Clement Escoffier</a>
 */
public class MyVerticle extends AbstractVerticle {


  @Override
  public void start() throws Exception {
    ServiceDiscovery discovery = ServiceDiscovery.create(vertx);
    EventBus eb = vertx.eventBus();

    eb.consumer("http-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-http-service")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            HttpClient client = reference.get();
            result.put("client", client.toString());
            result.put("direct", reference.get().toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("http-sugar", message -> {
      JsonObject result = new JsonObject();
      HttpEndpoint.getClient(discovery,
        record -> record.getName().equalsIgnoreCase("my-http-service")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            HttpClient client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("web-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-http-service")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            WebClient client = reference.getAs(WebClient.class);
            result.put("client", client.toString());
            result.put("direct", reference.get().toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("web-sugar", message -> {
      JsonObject result = new JsonObject();
      HttpEndpoint.getWebClient(discovery,
        record -> record.getName().equalsIgnoreCase("my-http-service")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            WebClient client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("service-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-service")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            HelloService client = reference.get();
            result.put("client", client.toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("service-sugar", message -> {
      JsonObject result = new JsonObject();
      EventBusService.getServiceProxy(discovery,
        record -> record.getName().equalsIgnoreCase("my-service"),
        HelloService.class).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            HelloService client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("ds-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-data-source")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            JDBCClient client = reference.get();
            result.put("client", client.toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("ds-sugar", message -> {
      JsonObject result = new JsonObject();
      JDBCDataSource.getJDBCClient(discovery, record -> record.getName().equalsIgnoreCase("my-data-source")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            JDBCClient client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("redis-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-redis-data-source")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            Redis client = reference.get();
            result.put("client", client.toString());
            result.put("direct", reference.get().toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("redis-sugar", message -> {
      JsonObject result = new JsonObject();
      RedisDataSource.getRedisClient(discovery, record -> record.getName().equalsIgnoreCase("my-redis-data-source")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            Redis client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("mongo-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-mongo-data-source")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            MongoClient client = reference.get();
            result.put("client", client.toString());
            result.put("direct", reference.get().toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("mongo-sugar", message -> {
      JsonObject result = new JsonObject();
      MongoDataSource.getMongoClient(discovery,
        record -> record.getName().equalsIgnoreCase("my-mongo-data-source")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            MongoClient client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });

    eb.consumer("source1-ref", message ->
      discovery.getRecord(rec -> rec.getName().equalsIgnoreCase("my-message-source-1")).onComplete(ar -> {
        if (ar.failed()) {
          message.reply("FAIL - No service");
        } else {
          JsonObject result = new JsonObject();
          ServiceReference reference = discovery.getReference(ar.result());
          if (reference == null) {
            message.reply("FAIL - reference is null");
          } else {
            MessageConsumer client = reference.get();
            result.put("client", client.toString());
            result.put("direct", reference.get().toString());
            reference.release();
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        }
      }));

    eb.consumer("source1-sugar", message -> {
      JsonObject result = new JsonObject();
      MessageSource.<String>getConsumer(discovery, record -> record.getName().equalsIgnoreCase("my-message-source-1")).onComplete(
        ar -> {
          if (ar.failed()) {
            message.reply("FAIL - no service");
          } else {
            MessageConsumer<String> client = ar.result();
            result.put("client", client.toString());
            ServiceDiscovery.releaseServiceObject(discovery, client);
            result.put("bindings", getBindings(discovery));
            message.reply(result);
          }
        });
    });


  }

  private synchronized JsonArray getBindings(ServiceDiscovery discovery) {
    JsonArray array = new JsonArray();
    for (ServiceReference ref : discovery.bindings()) {
      array.add(ref.toString());
    }
    return array;
  }
}
